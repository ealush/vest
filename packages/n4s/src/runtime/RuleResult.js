import hasOwnProperty from 'hasOwnProperty';

import { isBoolean } from 'isBoolean';
import { isEmpty } from 'isEmpty';
import { isNull } from 'isNull';
import { isUndefined } from 'isUndefined';
import { HAS_WARNINGS, HAS_ERRORS } from 'sharedKeys';

/**
 * Stores a rule result in an easy to inspect and manipulate structure.
 *
 * @param {boolean|RuleResult} ruleRunResult
 */
export default function RuleResult(ruleRunResult) {
  if (isUndefined(ruleRunResult)) {
    return;
  }

  if (isBoolean(ruleRunResult)) {
    this.setFailed(!ruleRunResult);
  } else {
    this.extend(ruleRunResult);
  }
}

/**
 * Determines whether a given value is a RuleResult instance
 * @param {*} res
 * @return {boolean}
 */
RuleResult.is = function (res) {
  return res instanceof RuleResult;
};

/**
 * Marks the current result object as an array
 */
RuleResult.prototype.asArray = function () {
  this.isArray = true;
  return this;
};

/**
 * @param {string} key
 * @param {value} value
 * @return {RuleResult} current instance
 */
RuleResult.prototype.setAttribute = function (key, value) {
  this[key] = value;
  return this;
};

/**
 * @param {boolean} failed
 * @return {RuleResult} current instance
 */
RuleResult.prototype.setFailed = function (failed) {
  this.setAttribute(this.warn ? HAS_WARNINGS : HAS_ERRORS, failed);
  return this.setAttribute('failed', failed);
};

/**
 * Adds a nested result object
 *
 * @param {string} key
 * @param {RuleResult} child
 */
RuleResult.prototype.setChild = function (key, child) {
  if (isNull(child)) {
    return null;
  }

  const isWarning =
    this[HAS_WARNINGS] || child[HAS_WARNINGS] || child.warn || this.warn;
  this.setAttribute(HAS_WARNINGS, (isWarning && child.failed) || false);
  this.setAttribute(
    HAS_ERRORS,
    this[HAS_ERRORS] ||
      child[HAS_ERRORS] ||
      (!isWarning && child.failed) ||
      false
  );
  this.setFailed(this.failed || child.failed);
  this.children = this.children || {};
  this.children[key] = child;
  return child;
};

/**
 * Retrieves a child by its key
 *
 * @param {string} key
 * @return {RuleResult|undefined}
 */
RuleResult.prototype.getChild = function (key) {
  return (this.children || {})[key];
};

/**
 * Extends current instance with a new provided result
 * @param {Boolean|RuleResult} newRes
 */
RuleResult.prototype.extend = function (newRes) {
  if (isNull(newRes)) {
    return this;
  }

  const res = RuleResult.is(newRes)
    ? newRes
    : new RuleResult().setAttribute('warn', !!this.warn).setFailed(!newRes);

  const failed = this.failed || res.failed;

  const children = mergeChildren(res, this).children;

  Object.assign(this, res);
  if (!isEmpty(children)) {
    this.children = children;
  }

  this.setFailed(failed);
  this.setAttribute(HAS_WARNINGS, !!(this[HAS_WARNINGS] || res[HAS_WARNINGS]));
  this.setAttribute(HAS_ERRORS, !!(this[HAS_ERRORS] || res[HAS_ERRORS]));
};

Object.defineProperty(RuleResult.prototype, 'pass', {
  get() {
    return !this.failed;
  },
});

/**
 * Deeply merge the nested children of compound rules
 *
 * @param {?RuleResult} base
 * @param {?RuleResult} add
 * @return {RuleResult}
 */
function mergeChildren(base, add) {
  const isRuleResultBase = RuleResult.is(base);
  const isRuleResultAdd = RuleResult.is(add);

  // If both base and add are result objects
  if (isRuleResultBase && isRuleResultAdd) {
    // Use failed if either is failing
    base.setFailed(base.failed || add.failed);

    // If neither has a children object, or the children object is
    if (isEmpty(base.children) && isEmpty(add.children)) {
      return base;
    }
    // If both have a children object
    if (base.children && add.children) {
      // Merge all the "right side" children back to base
      for (const key in base.children) {
        mergeChildren(base.children[key], add.children[key]);
      }

      // If a child exists in "add" but not in "base", just copy the child as is
      for (const key in add.children) {
        if (!hasOwnProperty(base.children, key)) {
          base.setChild(key, add.children[key]);
        }
      }

      // Return the modified base object
      return base;

      // If base has no children (but add does)
    } else if (!base.children) {
      // Use add's children
      base.children = add.children;

      // If add has no children
    } else if (!add.children) {
      // return base as is
      return base;
    }
    // If only base is `RuleResult`
  } else if (isRuleResultBase) {
    // Return base as is
    return base;

    // If only add is RuleResult
  } else if (isRuleResultAdd) {
    // Return add as is
    return add;
  }

  // That's a weird case. Let's fail. Very unlikely.
  return new RuleResult(false);
}
