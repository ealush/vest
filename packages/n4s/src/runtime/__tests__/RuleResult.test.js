import RuleResult from 'RuleResult';

describe('RuleResult', () => {
  describe('RuleResult constructor', () => {
    describe('When value is boolean', () => {
      it('Should instantiate with a matching `failed` value', () => {
        expect(new RuleResult(false).failed).toBe(true);
        expect(new RuleResult(true).failed).toBe(false);
        expect(new RuleResult(false).pass).toBe(false);
        expect(new RuleResult(true).pass).toBe(true);
      });
    });

    describe('When value is a RuleResult instance', () => {
      it('Should instantiate a new RuleResult instance with extended properties', () => {
        const prev = new RuleResult(false);
        prev.setChild('exampleChild', new RuleResult(true));

        const next = new RuleResult(prev);
        expect(next.failed).toBe(true);
        expect(next.children.exampleChild).toBe(prev.children.exampleChild);
        expect(next).not.toBe(prev);
        expect(next).toEqual(prev);
      });
    });
  });

  describe('RuleResult.is', () => {
    it('Should return true when value is a RuleResult instance', () => {
      expect(RuleResult.is(new RuleResult(true))).toBe(true);
      expect(RuleResult.is(new RuleResult(false))).toBe(true);
    });

    it('Should return false when value is not a RuleResult instance', () => {
      expect(RuleResult.is(true)).toBe(false);
      expect(RuleResult.is('Hello!')).toBe(false);
    });
  });

  describe('RuleResult.asArray', () => {
    it('Should mark with isArray: true', () => {
      expect(new RuleResult(true).isArray).toBeFalsy();
      const rr = new RuleResult(true).asArray();
      expect(rr.isArray).toBe(true);
    });
  });

  describe('RuleResult.setAttribute', () => {
    it('should set provided key/value pair', () => {
      const RR = new RuleResult(true);
      expect(RR.example).toBeUndefined();
      RR.setAttribute('example', 'YES!');
      expect(RR.example).toBe('YES!');
    });
  });

  describe('RuleResult.setChild', () => {
    it('Should create a children property if one does not exist', () => {
      const RR = new RuleResult(true);
      expect(RR.children).toBeUndefined();
      RR.setChild('example', new RuleResult(true));
      expect(RR.children).toBeDefined();
    });

    it('Should set the RuleResult as failed if the child is failed', () => {
      const RR = new RuleResult(true);
      expect(RR.pass).toBe(true);
      RR.setChild('example', new RuleResult(new RuleResult(false)));
      expect(RR.pass).toBe(false);
    });

    it('Should set the provided key with the provided child object', () => {
      const RR = new RuleResult(true);
      const child = new RuleResult(true);
      RR.setChild('example', child);
      expect(RR.children.example).toBe(child);
    });
  });

  describe('RuleResult.getChild', () => {
    it('Should return child if exists', () => {
      const RR = new RuleResult(true);
      const child = new RuleResult(true);
      RR.setChild('child', child);
      expect(RR.getChild('child')).toBe(child);
    });

    it('Should return undefined when RuleResult has no children', () => {
      const RR = new RuleResult(true);
      expect(RR.getChild('child')).toBeUndefined();
    });
  });

  describe('RuleResult.pass', () => {
    it('Should negate "failed" property', () => {
      const RR = new RuleResult(true);
      expect(RR.failed).toBe(false);
      expect(RR.pass).toBe(true);
      RR.setFailed(true);
      expect(RR.failed).toBe(true);
      expect(RR.pass).toBe(false);
    });
  });

  describe('RuleResult.extend', () => {
    it('Should use most severe failed state', () => {
      let left = new RuleResult(true);
      let right = new RuleResult(false);
      left.extend(right);
      expect(left.failed).toBe(true);
      left = new RuleResult(false);
      right = new RuleResult(true);
      left.extend(right);
      expect(left.failed).toBe(true);
    });

    it('Should match failed state even when a few layers deep', () => {
      const left = new RuleResult(true);
      left.setChild(
        'l1',
        new RuleResult(true).setChild('l2', new RuleResult(true))
      );
      const right = new RuleResult(true);
      right.setChild(
        'l1',
        new RuleResult(true).setChild('l2', new RuleResult(false))
      );
      expect(left.pass).toBe(true);
      expect(right.pass).toBe(false);
      left.extend(right);
      expect(left.pass).toBe(false);
    });
    it('Should use most severe failed value', () => {
      let left = new RuleResult(true);
      let right = new RuleResult(true);

      left.setChild('example', new RuleResult(true));
      right.setChild('example', new RuleResult(false));

      left.extend(right);
      expect(left.getChild('example').failed).toBe(true);
      left = new RuleResult(true);
      right = new RuleResult(true);
      right.setChild('example', new RuleResult(true));
      left.setChild('example', new RuleResult(false));
      left.extend(right);
      expect(left.getChild('example').failed).toBe(true);
      expect(left).toMatchSnapshot();
    });

    it('Should merge children keys', () => {
      const left = new RuleResult(true);
      const right = new RuleResult(true);
      left.setChild('a', new RuleResult(true));
      right.setChild('b', new RuleResult(true));
      right.setChild('c', new RuleResult(true));
      left.extend(right);
      expect(left.children.b).toBe(right.children.b);
      expect(left.children.c).toBe(right.children.c);

      expect(left).toMatchSnapshot();
    });

    it('Should deep merge children keys', () => {
      const l = new RuleResult(true);
      const l_a = new RuleResult(true);
      const l_a_1 = new RuleResult(true);
      const l_a_1_x = new RuleResult(true);
      l_a_1.setChild('x', l_a_1_x);
      l_a.setChild('1', l_a_1);
      l.setChild('a', l_a);

      const r = new RuleResult(true);
      const r_a = new RuleResult(true);
      const r_a_1 = new RuleResult(true);
      const r_a_1_x = new RuleResult(false);
      r_a_1.setChild('x', r_a_1_x);
      r_a.setChild('1', r_a_1);
      r.setChild('a', r_a);

      expect(l.pass).toBe(true);
      l.extend(r);
      expect(l.pass).toBe(false);

      expect(l).toMatchSnapshot();
    });
  });
});
