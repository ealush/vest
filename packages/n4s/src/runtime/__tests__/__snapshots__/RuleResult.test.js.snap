// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuleResult RuleResult.extend Should deep merge children keys 1`] = `
RuleResult {
  "children": Object {
    "a": RuleResult {
      "children": Object {
        "1": RuleResult {
          "children": Object {
            "x": RuleResult {
              "failed": true,
              "hasErrors": true,
            },
          },
          "failed": true,
          "hasErrors": true,
          "hasWarnings": false,
        },
      },
      "failed": true,
      "hasErrors": true,
      "hasWarnings": false,
    },
  },
  "failed": true,
  "hasErrors": true,
  "hasWarnings": false,
}
`;

exports[`RuleResult RuleResult.extend Should merge children keys 1`] = `
RuleResult {
  "children": Object {
    "a": RuleResult {
      "failed": false,
      "hasErrors": false,
    },
    "b": RuleResult {
      "failed": false,
      "hasErrors": false,
    },
    "c": RuleResult {
      "failed": false,
      "hasErrors": false,
    },
  },
  "failed": false,
  "hasErrors": false,
  "hasWarnings": false,
}
`;

exports[`RuleResult RuleResult.extend Should use most severe failed value 1`] = `
RuleResult {
  "children": Object {
    "example": RuleResult {
      "failed": true,
      "hasErrors": true,
    },
  },
  "failed": true,
  "hasErrors": true,
  "hasWarnings": false,
}
`;
